src/server/server/handlers.py

```diff
@@ -1,69 +1,153 @@
+            self.response.write(response)
+
+
+class RegisterHandler(BaseHandler):
+
+    def get(self):
+        user = self.get_connected_user()
+        if user is None:
+            response = jinjaenvironment.render_template(html_file='register.html', attributes_dict={})
+            self.response.write(response)
+        else:
+            self.redirect('/')
 
     def post(self):
-        user = self._get_or_create_user()
-        feature_vector = self._get_or_create_features(user)
-        labels = dal.get_features(feature_vector.feature_set_id)
-        user_docs = dal.get_user_docs(user)
-        links = [Link(link=user_doc.document.url, text=user_doc.document.title) for user_doc in user_docs]
-        features = [Feature(label, value) for (label, value) in zip(labels, feature_vector.vector)]
-
-        template_values = {
-            'email': user.email,
-            'links': links,
-            'features': features,
-            'feature_set_id': feature_vector.feature_set_id
-        }
-        response = jinjaenvironment.render_template(html_file='index.html', attributes_dict=template_values)
-
-        self.response.write(response)
+        connected_user = self.get_connected_user()
+        if connected_user is None:
+            email = self.request.get('email')
+            password = self.request.get('password')
+            interests = self.request.get('interests')
+
+            user = dal.get_user(email)
+            if user is None:
+                user = struct.User.make_from_scratch(email, helpers.PasswordHelpers.hash_password(password), interests)
+                dal.save_user(user)
+
+                # Create an empty feature vector for the newly created user
+                features_set = dal.get_features(dal.REF_FEATURE_SET)
+                feature_vector = struct.FeatureVector.make_from_scratch([0]*len(features_set), dal.REF_FEATURE_SET)
+                dal.save_user_feature_vector(user, feature_vector)
+
+                self.redirect('/login')
+            else:
+                template_value = {'error_message': 'This account already exists'}
+                response = jinjaenvironment.render_template(html_file='register.html', attributes_dict=template_value)
+                self.response.write(response)
+        else:
+            self.redirect('/')
+
+
+class DisconnectHandler(BaseHandler):
+
+    def get(self):
+        user = self.get_connected_user()
+        if user is not None:
+            self.unset_connected_user()
+        self.redirect('/login')
+
+
+class HomeHandler(BaseHandler):
+
+    def get(self):
+        user = self.get_connected_user()
+        if user is not None:
+            user_docs = dal.get_user_docs(user)
+            links = [Link(key=user_doc.document._db_key.urlsafe(), # pylint: disable=protected-access
+                          text=user_doc.document.title) for user_doc in user_docs]
```

Comment:
To not break datastore encapsulation (not use _db_key), I think we should have a field "id_urlsafe" that is set at construction time from dal.